### PR TITLE
Getting a list of pages (for sitemaps, etc.)

### PR DESCRIPTION
**Pull request by bentoncreation**
Here's my take on getting a list of High Voltage pages for building sitemaps, based on this issue: https://github.com/thoughtbot/high_voltage/issues/123

If you're using the sitemap_generator gem, you could do something like this in your config:

```
HighVoltage.page_ids.each do |page|
  add page, changefreq: 'monthly'
end
```

### CODE REVIEW

**Comment by bentoncreation** [comment]
I fixed the empty line, new line and "line too long" issues from Hound.

Should I make the double-quoted string changes? Seemed like most of the repo is using single quotes.

File: lib/high_voltage/configuration.rb [hunk 1]
```diff
@@ -14,7 +14,8 @@ module Configuration
       :routes,
     )
 
-    attr_reader :action_caching, :action_caching_layout, :page_caching
+    attr_reader :action_caching, :action_caching_layout, :page_caching,
+                :page_ids
```
Comment (dgalarza) [inline lib/high_voltage/configuration.rb]: Do we need this `attr_reader` if we are defining `page_ids` on line 43?

Comment (bentoncreation) [inline lib/high_voltage/configuration.rb]: Ah! You're right, not sure what I was thinking there.

**Comment by dgalarza** [comment]
Thanks for the pr @bentoncreation. 

> Should I make the double-quoted string changes? Seemed like most of the repo is using single quotes

We're slowly in the process of moving towards double quotes. Instead of doing this via one big sweep we're just making changes as we go through. As such, any places that are touched we recommend moving to double quote usage.

Some other thoughts:

What do you think about extracting another class to represent a page (or view)? That way instead of having a few private methods in the `PageCollector` class that need the same argument passed around, it could be contained within this new object.

Something along the lines of:

```ruby
if page.valid?
  page_paths.push page.path
end
```

**Comment by bentoncreation** [comment]
Sure thing @dgalarza ! Thanks for the review.

Okay, I'll follow Hound on double quotes, etc. (Although, I think it's wrong about 'detect' over 'find'. It seems to be thinking it's Enumerable, not the Find module.)

A Page class makes sense. How would that relate to the PageFinder class? Are they overlapping at all? I want to make sure I get the right abstraction.

File: lib/high_voltage/page_collector.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    attr_reader :page_ids, :content_path
+
+    def initialize(content_path)
+      @content_path = content_path.to_s
+    end
+
+    def page_ids
+      page_ids = []
+
+      Find.find(content_path) do |f|
```
Comment (dgalarza) [inline lib/high_voltage/page_collector.rb]: @bentoncreation good call. I think your right that it's thinking of enumerable over `Find`

**Comment by dgalarza** [comment]
I don't think that the new class would overlap with the `PageFinder` class since `PageFinder` is specifically responsible for finding a given static page based off an ID.

**Comment by bentoncreation** [comment]
Ok, cool. Also, I noticed in your sample code, you changed page_ids to page_paths. The original issue suggested page_ids as the interface...would you rather go with page_paths?

**Comment by dgalarza** [comment]
Ah good question. I named it `page_paths` in my examples since it looked like you were collecting the path to a page with `url_path`. It feels like a better name imo however I don't feel to strongly about it if you feel otherwise.

**Comment by bentoncreation** [comment]
Does the following make sense to you?

It seems like we're trying to make a distinction between file paths and URL paths. They could be different, since you can override the PageFinder with the Rot13PageFinder. To keep that distinction, we're calling the URL path a page_id.

PageFinder says: give me a page_id and I'll find the file path.

PageCollector is saying: give me a base file path and I'll return the page_ids for all the files I find there.

**Comment by dgalarza** [comment]
Makes sense to me, :+1:

**Comment by bentoncreation** [comment]
Great, thanks!

**Comment by bentoncreation** [comment]
Hey @dgalarza, I've broken out the Page class we talked about and I think I have things ready for you to review. Let me know how this looks when you get the chance. Thanks!

File: spec/high_voltage/page_spec.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  it "produces the id for a page" do
+    expect(page(full_file_path("exists.html.erb")).id).to eq "exists"
+  end
+
+  it "produces the id for a page in a subdirectory" do
+    expect(page(full_file_path("dir/nested.html.erb")).id).to eq "dir/nested"
+  end
+
+  it "is valid for a page" do
+    expect(page(full_file_path("exists.html.erb")).valid?).to be true
```
Comment (dgalarza) [inline spec/high_voltage/page_spec.rb]: Consider using:

```ruby
expect(page(full_file_path("exists.html.erb"))).to be_valid
```

Comment (dgalarza) [inline spec/high_voltage/page_spec.rb]: It's also a little hard to follow along with the number of nested method calls and their opening / closing parens. What do you think about making it easier to read with something like:

```ruby
page = page(full_file_path("exists.html.erb"))

expect(page).to be_valid
```

throughout.

File: lib/high_voltage/page_collector.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    attr_reader :content_path
+
+    def initialize(content_path)
+      @content_path = content_path.to_s
+    end
+
+    def page_ids
+      page_ids = []
+
+      Find.find(content_path) do |file_path|
```
Comment (dgalarza) [inline lib/high_voltage/page_collector.rb]: This method is getting a bit long and seems to be responsible for quite a bit. What do you think about extracting a private method for collecting Page objects and then using a select on that collection for valid ones.

Something like:

```ruby
def page_ids
  pages.select(&:valid?).map(&:id)
end

private

def pages
  Find.find(content_path).map { |file_path| HighVoltage::Page.new(content_path, file_path) }
end
```

File: .gitignore [hunk 5]
```diff
@@ -4,3 +4,4 @@ pkg/
 *.swp
 *.gem
 gemfiles/*.lock
+spec/fixtures/log/*.log
```
Comment (dgalarza) [inline .gitignore]: Missing a new line here.

File: spec/fixtures/app/views/pages/_partial.html.erb [hunk 6]
```diff
@@ -0,0 +1 @@
+partial
```
Comment (dgalarza) [inline spec/fixtures/app/views/pages/_partial.html.erb]: Missing a new line here as well.

File: spec/fixtures/app/views/pages/text.txt.erb [hunk 7]
```diff
@@ -0,0 +1 @@
+text page
```
Comment (dgalarza) [inline spec/fixtures/app/views/pages/text.txt.erb]: Missing a new line here as well.

**Comment by dgalarza** [comment]
@bentoncreation this is looking really good. Just a few more stylistic tweaks and I think this will be ready to merge soon. Thanks for keeping up with it @bentoncreation.

**Comment by bentoncreation** [comment]
@dgalarza Hey, thanks for the guidance/patience! :) I'm going to try to make your recommended fixes this morning.

File: lib/high_voltage/page_collector.rb (hunk 4 above)
Comment (bentoncreation) [inline lib/high_voltage/page_collector.rb]: Ah yes, much nicer!

**Comment by dgalarza** [comment]
Awesome stuff @bentoncreation. Looks good to me. Would you mind squashing to prep this for a merge? Otherwise I can go ahead and do so.

Also, it might be useful to add something in the README about this and it's applicable usage.

**Comment by aguynamedben** [comment]
:+1: nice work @bentoncreation. I'm not familiar with this codebase but I CRed and LGTM. Nice!

**Comment by bentoncreation** [comment]
Thanks, @dgalarza! I'll work on a README update shortly.

Thanks, @aguynamedben!

**Comment by bentoncreation** [comment]
Hey @dgalarza, can you let me know if the README update looks ok?

**Comment by dgalarza** [comment]
Looks good to me @bentoncreation

**Comment by bentoncreation** [comment]
@dgalarza Okay, I think I squashed it right. That was my first time doing an interactive rebase. :) Let me know if anything else is needed.

**Comment by dgalarza** [comment]
Thanks again for the hard work on this @bentoncreation. I'll put out a release with this in it later in the week. :+1:

**Comment by bentoncreation** [comment]
Sounds good, thanks!
